### ISSUE TITLE
Potential bug that halts sync if obsolete order data is present in HPOS when legacy tables are authoritative and sync enabled

### ISSUE THREAD

**Issue by kaushikasomaiya** [woothemes/woocommerce#52837]
### Prerequisites

- [x] I have carried out troubleshooting steps and I believe I have found a bug.
- [x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

#8933505-zen

Merchant reported they are unable to sync orders to HPOS. They have legacy tables enabled currently with sync enabled.

During investigation I've found the following:

- When a post is deleted from legacy tables, ideally the synced order on HPOS tables should also get deleted 
- In merchant's case, something prevented the order from being deleted from the HPOS tables when it was deleted from WP Admin
- This causes the below fatal error while syncing 

> action failed via Async Request: get_class(): Argument #1 ($object) must be of type object, null given

https://github.com/woocommerce/woocommerce/blob/7f32086ae0e9e4370fc5a111e5d6c9d70785e96f/plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php#L1417

- This is because of obsolete order present in HPOS with the meta key `_deleted_from` and value `posts_table` I think this causes get_post_orders_for_ids() to pick it some how 

### Expected behavior

Ideally, regardless of the obsolete order, sync should not be affected. 

### Actual behavior

NA

### Steps to reproduce

I don't have steps to replicate it. 

### WordPress Environment

NA

### Isolating the problem

- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/products/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

**Comment by nicdwilson** [comment]
p1729654529944829-slack-C7U3Y3VMY and 8854597-zd-a8c

**Comment by kaushikasomaiya** [comment]
I think I have the steps to replicate:

1. Enable legacy as authoritative and no sync
2. Place a couple of orders 
3. Enable sync but let legacy tables still authoritative; let the orders sync to HPOS
4. Once everything is synced, disable sync
5. Trash and delete one order from WP Admin
6. _deleted_from meta will be inserted in the HPOS meta table 
7. Enable sync again and try to sync to replicate the error.

**Comment by Poocey** [comment]
Any update on this?

**Comment by kaushikasomaiya** [comment]
@Poocey Are you trying to upgrade to HPOS? There's a workaround available for this issue.

**Comment by Poocey** [comment]
> [@Poocey](https://github.com/Poocey) Are you trying to upgrade to HPOS? There's a workaround available for this issue.

Yes. I'm stuck on 226 orders pending. What's the workaround?

**Comment by kaushikasomaiya** [comment]
Please do the following (Assume you're currently on legacy storage )

- Disable the sync/compatibility mode checkmark and save settings (Under Woo > Settings > Advanced > Features)
- Go to Woo > Status > Tools area and delete all Custom Order Tables (There is a tool for it)
- Go to Woo > Settings > Advanced > Features and enable the sync checkmark, and save.

This should help restart the sync.

**Comment by Poocey** [comment]
> Please do the following (Assume you're currently on legacy storage )
> 
>     * Disable the sync/compatibility mode checkmark and save settings (Under Woo > Settings > Advanced > Features)
> 
>     * Go to Woo > Status > Tools area and delete all Custom Order Tables (There is a tool for it)
> 
>     * Go to Woo > Settings > Advanced > Features and enable the sync checkmark, and save.
> 
> 
> This should help restart the sync.

We have well over 1M orders. I'll have to resync all of them to fix this?

**Comment by kaushikasomaiya** [comment]
@Poocey Thanks for the additional context, in that case we suggest opening a ticket so our support team can check further https://woocommerce.com/my-account/contact-support/ Please mention this issue link in the description. You'll need to create a free WooCommerce.com account before you can submit a ticket.

**Comment by Poocey** [comment]
> [@Poocey](https://github.com/Poocey) Thanks for the additional context, in that case we suggest opening a ticket so our support team can check further https://woocommerce.com/my-account/contact-support/ Please mention this issue link in the description. You'll need to create a free WooCommerce.com account before you can submit a ticket.

There are currently 106 total: 99 ID_TYPE_MISSING_IN_POSTS_TABLE and 7 ID_TYPE_DIFFERENT_UPDATE_DATE

Is this possible to fix manually?

**Comment by kaushikasomaiya** [comment]
Do you see any results for 
```
SELECT * FROM `wp_wc_orders_meta` WHERE `meta_key` LIKE '%_deleted_from%';
```
These can be potentially obsolete orders deleted from legacy tables. I am also assuming you're currently on legacy storage and sync is enabled.

**Comment by Poocey** [comment]
> Do you see any results for
> 
> ```
> SELECT * FROM `wp_wc_orders_meta` WHERE `meta_key` LIKE '%_deleted_from%';
> ```
> 
> These can be potentially obsolete orders deleted from legacy tables. I am also assuming you're currently on legacy storage and sync is enabled.

Yes. I'm getting the same results for this query as I do when I run the DataAnalyzer query w/ ID_TYPE_MISSING_IN_POSTS_TABLE type (99 results).

**Comment by kaushikasomaiya** [comment]
Next you can analyze what orders they were and if you need them - if not you can back them up somewhere and delete from the HPOS DB tables. 

```
SELECT * FROM `wp_wc_orders_meta` WHERE `order_id` IN (SELECT `order_id` FROM `wp_wc_orders_meta` WHERE `meta_key` LIKE '%_deleted_from%')
```

```
SELECT * FROM `wp_wc_orders` WHERE `id` IN (SELECT `order_id` FROM `wp_wc_orders_meta` WHERE `meta_key` LIKE '%_deleted_from%')
```

```
SELECT * FROM `wp_wc_order_addresses` WHERE `order_id` IN (SELECT `order_id` FROM `wp_wc_orders_meta` WHERE `meta_key` LIKE '%_deleted_from%')
```

```
SELECT * FROM `wp_wc_order_operational_data` WHERE `order_id` IN (SELECT `order_id` FROM `wp_wc_orders_meta` WHERE `meta_key` LIKE '%_deleted_from%')
```

Note that this should be done if your batch actions are failing with "`get_class(): Argument 1 ($object) must be of type object, null given`" - if not, it might be a different sync issue that needs investigation.

**Comment by Poocey** [comment]
@kaushikasomaiya If I need them, how can I bypass the get_class() error? Yet if I don't need, would I just wpdb->delete() these from wp_wc_orders table (or all the tables to satisfy HPOS)?

Also, for the ID_TYPE_DIFFERENT_UPDATE_DATE  issue: Would I just match the last_updated dates for the orders between wp_posts and wc_wc_orders?

**Comment by kaushikasomaiya** [comment]
@Poocey if you need them - it might be tricky to bring them back to legacy tables - one option can be of course delete the _deleted_from meta entry so they sync as normal orders however whether there would be side effects of it during sync is something that needs investigation. Easier option is you can use the four commands I sent above to study which orders they were, and just use export command in a tool like phpMyAdmin to backup them and delete. It is safer to delete them from all four tables.

> Also, for the ID_TYPE_DIFFERENT_UPDATE_DATE issue: Would I just match the last_updated dates for the orders between wp_posts and wc_wc_orders?

I believe this will be taken care of by Sync once we're able to bypass the get_class error; which is by deleting the obsolete orders from all HPOS tables.
